api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,145 @@
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context SetValues(T &values) noexcept
+  {
+    Context context = Context(values);
+    nostd::shared_ptr<DataList> last_node;
+    for (nostd::shared_ptr<DataList> data = context.head_; data != nullptr; data = data->next_)
+    {
+      last_node = data;
+    }
+    last_node->next_ = head_;
+    return context;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data. It attaches the
+  // exisiting list to the end of the new list.
+  Context SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context context      = Context(key, value);
+    context.head_->next_ = head_;
+    return context;
+  }
+
+  // Returns the value associated with the passed in key.
+  context::ContextValue GetValue(nostd::string_view key)
+  {
+    for (nostd::shared_ptr<DataList> data = head_; data != nullptr; data = data->next_)
+    {
+      if (key.size() == data->key_length_)
+      {
+        if (memcmp(key.data(), data->key_, data->key_length_) == 0)
+        {
+          return data->value_;
+        }
+      }
+    }
+    return "";
+  }
+
+private:
+  Context() = default;
+
+  // A linked list to contain the keys and values of this context node
+  class DataList
+  {
+  public:
+    nostd::shared_ptr<DataList> next_;
+
+    char *key_;
+
+    size_t key_length_;
+
+    ContextValue value_;
+
+    DataList() { next_ = nullptr; }
+
+    // Builds a data list off of a key and value iterable and returns the head
+    template <class T,
+              nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+    DataList(const T &keys_and_vals)
+    {
+      auto iter = std::begin(keys_and_vals);
+      // Create list head
+      key_        = new char[nostd::string_view(iter->first).size()];
+      key_length_ = nostd::string_view(iter->first).size();
+      memcpy(key_, nostd::string_view(iter->first).data(),
+             nostd::string_view(iter->first).size() * sizeof(char));
+      value_ = iter->second;
+      next_  = nostd::shared_ptr<DataList>{nullptr};
+      ++iter;
+
+      DataList *previous_node = this;
+      // Iterate over the keys and values iterable and add nodes
+      for (; iter != std::end(keys_and_vals); ++iter)
+      {
+        nostd::shared_ptr<DataList> node = nostd::shared_ptr<DataList>{new DataList()};
+        node->next_                      = nostd::shared_ptr<DataList>{nullptr};
+        node->key_                       = new char[nostd::string_view(iter->first).size()];
+        node->key_length_                = nostd::string_view(iter->first).size();
+        memcpy(node->key_, nostd::string_view(iter->first).data(),
+               nostd::string_view(iter->first).size() * sizeof(char));
+        node->value_         = iter->second;
+        previous_node->next_ = node;
+        previous_node        = &*node;
```

Comment:
You can make this a `previous_node = node`.

Reply:
My goal he was to avoid making previous_node a shared_ptr, because I don't have access to the other shared_ptr pointing to this object at the time, and creating another would cause the node to destruct too early.

Reply:
*here